File: writebook/app/models/p2u/planner.rb
module P2u
  # Declarative plan/apply. Compares a desired manifest with an existing deck
  # (matched by stable slide `p2u_id`, with a positional fallback for legacy
  # decks) and produces an ordered list of create/update/delete/reorder actions.
  class Planner
    class Action
      ATTRIBUTES = %i[action slide leaf_id changes slides].freeze
      attr_reader(*ATTRIBUTES)

      def initialize(action:, slide: nil, leaf_id: nil, changes: nil, slides: nil)
        @action = action.to_s
        @slide = slide
        @leaf_id = leaf_id
        @changes = changes
        @slides = slides
      end

      def to_h = ATTRIBUTES.index_with { |attribute| send(attribute) }.compact
    end

    class Result
      attr_reader :diagnostics, :actions, :book

      def initialize(diagnostics:, actions:, book: nil)
        @diagnostics = diagnostics
        @actions = actions
        @book = book
      end

      def valid? = diagnostics.none?(&:error?)
      def summary = actions.group_by(&:action).transform_values(&:size)

      def to_h
        {
          valid: valid?,
          deck: book && { id: book.id, title: book.title, slug: book.slug },
          summary: summary,
          actions: actions.map(&:to_h),
          diagnostics: diagnostics.map(&:to_h)
        }.compact
      end
    end

    def initialize(manifest, book:)
      @manifest = manifest.is_a?(Manifest) ? manifest : Manifest.new(manifest)
      @book = book
      @desired = Emitter.new(@manifest).slides.index_by { |slide| slide[:id] }
    end

    def plan
      diagnostics = Validator.new(@manifest).validate
      actions = diagnostics.any?(&:error?) ? [] : build_actions

      Result.new(diagnostics: diagnostics, actions: actions, book: @book)
    end

    def apply(user:)
      result = plan
      return result unless result.valid?

      @matches = matches

      @book.transaction do
        mapping = matches.dup
        result.actions.each do |action|
          case action.action
          when "create" then mapping[action.slide] = create_slide(action, user)
          when "update" then mapping[action.slide] = update_slide(action)
          when "delete" then delete_slide(action)
          end
        end

        reorder(mapping) if result.actions.any? { |action| action.action == "reorder" }
      end

      result
    end

    private
      def existing_leaves
        @existing_leaves ||= @book.leaves.active.with_leafables.positioned.to_a
      end

      def matches
        @matches ||= compute_matches
      end

      def compute_matches
        desired = @manifest.slides
        matched = {}

        existing_leaves.each do |leaf|
          next if leaf.p2u_id.blank?

          matched[leaf.p2u_id] = leaf if desired.any? { |slide| slide["id"] == leaf.p2u_id }
        end

        unmatched_desired = desired.reject { |slide| matched.key?(slide["id"]) }
        unmatched_existing = existing_leaves.reject { |leaf| matched.value?(leaf) }
        unmatched_desired.zip(unmatched_existing).each do |slide, leaf|
          matched[slide["id"]] = leaf if leaf
        end

        matched
      end

      def build_actions
        actions = []

        @manifest.slides.each do |slide|
          leaf = matches[slide["id"]]

          if leaf
            changes = changes_for(leaf, @desired.fetch(slide["id"]))
            actions << Action.new(action: "update", slide: slide["id"], leaf_id: leaf.id, changes: changes) if changes.any?
          else
            actions << Action.new(action: "create", slide: slide["id"])
          end
        end

        existing_leaves.each do |leaf|
          next if matches.value?(leaf)

          actions << Action.new(action: "delete", slide: leaf.p2u_id.presence || leaf.title, leaf_id: leaf.id)
        end

        if reorder_needed?
          actions << Action.new(action: "reorder", slides: @manifest.slides.map { |slide| slide["id"] })
        end

        actions
      end

      def reorder_needed?
        return true if @manifest.slides.any? { |slide| matches[slide["id"]].nil? }

        current = existing_leaves.select { |leaf| matches.value?(leaf) }.map(&:id)
        expected = @manifest.slides.filter_map { |slide| matches[slide["id"]]&.id }
        current != expected
      end

      def changes_for(leaf, desired)
        changes = {}
        compare(changes, "title", leaf.title, desired[:title])
        compare(changes, "notes", leaf.notes, desired[:notes])
        compare(changes, "sketch", leaf.sketch, desired[:sketch])
        compare(changes, "layout", current_layout(leaf), desired[:layout])
        compare_leafable(changes, leaf, desired)
        changes
      end

      def current_layout(leaf)
        leaf.layout.presence || Layouts.resolve(leaf.leafable_name)
      end

      def compare_leafable(changes, leaf, desired)
        case leaf.leafable_name.to_s
        when "page"
          compare(changes, "body", leaf.page.body.content.to_s, desired[:body])
        when "section"
          compare(changes, "body", leaf.section.body, desired[:body])
          compare(changes, "theme", leaf.section.theme, desired[:theme])
        when "picture"
          compare(changes, "caption", leaf.picture.caption, desired[:caption])
        when "typst"
          compare(changes, "source", leaf.typst.source, desired[:source])
        end
      end

      def compare(changes, field, current, desired)
        return if current.to_s.strip == desired.to_s.strip

        changes[field] = [ current, desired ]
      end

      def create_slide(action, user)
        DeckBuilder.new(user).add_slide(@book, @desired.fetch(action.slide))
      end

      def update_slide(action)
        leaf = @book.leaves.active.find(action.leaf_id)
        desired = @desired.fetch(action.slide)

        leaf.edit(leafable_params: leafable_params_for(leaf, desired), leaf_params: leaf_params_for(desired))
        leaf.reload
      end

      def delete_slide(action)
        @book.leaves.active.find(action.leaf_id).trashed!
      end

      def leaf_params_for(desired)
        {
          title: desired[:title],
          notes: desired[:notes],
          sketch: desired[:sketch],
          layout: desired[:layout],
          p2u_id: desired[:id]
        }
      end

      def leafable_params_for(leaf, desired)
        case leaf.leafable_name.to_s
        when "page" then { body: desired[:body].to_s }
        when "section" then { body: desired[:body], theme: desired[:theme] }
        when "typst" then { source: desired[:source].to_s }
        when "picture" then { caption: desired[:caption] }
        else {}
        end
      end

      # Move every slide to the end in manifest order, which leaves the deck in
      # the desired order.
      def reorder(mapping)
        count = @book.leaves.active.count

        @manifest.slides.each do |slide|
          leaf = mapping[slide["id"]] || @book.leaves.active.find_by(p2u_id: slide["id"])
          leaf&.move_to_position(count)
        end
      end
  end
end

File: cloudflare/src/compiler/planner.ts
// Declarative plan/apply. Compares a desired manifest with an existing deck
// (matched by stable slide `p2u_id`, with a positional fallback) and produces an
// ordered list of create/update/delete/reorder actions. Port of p2u/planner.rb.
//
// The diff is pure; `apply` is left to the store layer (D1) so this module stays
// testable without a database.

import { Manifest } from "./manifest";
import type { Slide } from "./manifest";
import { Diagnostic } from "./diagnostic";
import { Validator } from "./validator";
import { Emitter } from "./emitter";
import type { SlideAttributes } from "./emitter";
import { layoutKey, resolveLayout } from "./layouts";

export interface ExistingSlide {
  id: number;
  p2u_id: string | null;
  title: string | null;
  notes: string | null;
  sketch: boolean;
  layout: string | null;
  kind: string;
  body: string | null;
  source: string | null;
  caption: string | null;
  theme?: string | null;
}

export type ActionName = "create" | "update" | "delete" | "reorder";

export interface Action {
  action: ActionName;
  slide?: string;
  leaf_id?: number;
  changes?: Record<string, [unknown, unknown]>;
  slides?: string[];
}

export interface PlanResult {
  valid: boolean;
  deck: { id?: number; title?: string; slug?: string };
  summary: Record<string, number>;
  actions: Action[];
  diagnostics: Diagnostic[];
}

export class Planner {
  private readonly manifest: Manifest;
  private readonly existing: ExistingSlide[];
  private readonly desired: Record<string, SlideAttributes>;
  private matches: Record<string, ExistingSlide> = {};

  constructor(
    manifest: Manifest | unknown,
    existing: ExistingSlide[],
    deck: { id?: number; title?: string; slug?: string } = {},
  ) {
    this.manifest = manifest instanceof Manifest ? manifest : new Manifest(manifest);
    this.existing = existing;
    this.deck = deck;
    this.desired = Object.fromEntries(new Emitter(this.manifest).slides.map((slide) => [slide.id, slide]));
  }

  private readonly deck: { id?: number; title?: string; slug?: string };

  plan(): PlanResult {
    const diagnostics = Validator.validate(this.manifest);
    const actions = diagnostics.some((diagnostic) => diagnostic.isError) ? [] : this.buildActions();

    const summary: Record<string, number> = {};
    for (const action of actions) summary[action.action] = (summary[action.action] ?? 0) + 1;

    return {
      valid: !diagnostics.some((diagnostic) => diagnostic.isError),
      deck: this.deck,
      summary,
      actions,
      diagnostics,
    };
  }

  /** The ordered write plan the store layer should execute. */
  get writePlan(): Action[] {
    return this.plan().actions;
  }

  private computeMatches(): Record<string, ExistingSlide> {
    const desiredSlides = this.manifest.slides;
    const matched: Record<string, ExistingSlide> = {};

    for (const slide of this.existing) {
      if (!slide.p2u_id) continue;
      if (desiredSlides.some((desired) => desired.id === slide.p2u_id)) matched[slide.p2u_id] = slide;
    }

    const unmatchedDesired = desiredSlides.filter((slide) => !(slide.id in matched));
    const matchedIds = new Set(Object.values(matched).map((slide) => slide.id));
    const unmatchedExisting = this.existing.filter((slide) => !matchedIds.has(slide.id));

    unmatchedDesired.forEach((slide, index) => {
      const candidate = unmatchedExisting[index];
      if (candidate) matched[slide.id] = candidate;
    });

    return matched;
  }

  private buildActions(): Action[] {
    this.matches = this.computeMatches();
    const actions: Action[] = [];

    for (const slide of this.manifest.slides) {
      const existing = this.matches[slide.id];
      if (existing) {
        const changes = this.changesFor(existing, this.desired[slide.id]!);
        if (Object.keys(changes).length > 0) {
          actions.push({ action: "update", slide: slide.id, leaf_id: existing.id, changes });
        }
      } else {
        actions.push({ action: "create", slide: slide.id });
      }
    }

    const matchedIds = new Set(Object.values(this.matches).map((slide) => slide.id));
    for (const slide of this.existing) {
      if (matchedIds.has(slide.id)) continue;
      actions.push({ action: "delete", slide: slide.p2u_id ?? slide.title ?? undefined, leaf_id: slide.id });
    }

    if (this.reorderNeeded()) {
      actions.push({ action: "reorder", slides: this.manifest.slides.map((slide) => slide.id) });
    }

    return actions;
  }

  private reorderNeeded(): boolean {
    if (this.manifest.slides.some((slide) => !this.matches[slide.id])) return true;

    const matchedIds = new Set(Object.values(this.matches).map((slide) => slide.id));
    const current = this.existing.filter((slide) => matchedIds.has(slide.id)).map((slide) => slide.id);
    const expected = this.manifest.slides
      .map((slide) => this.matches[slide.id]?.id)
      .filter((id): id is number => id !== undefined);
    return current.join(",") !== expected.join(",");
  }

  private changesFor(existing: ExistingSlide, desired: SlideAttributes): Record<string, [unknown, unknown]> {
    const changes: Record<string, [unknown, unknown]> = {};
    compare(changes, "title", existing.title, desired.title);
    compare(changes, "notes", existing.notes, desired.notes);
    compare(changes, "sketch", existing.sketch, desired.sketch);
    compare(changes, "layout", existing.layout ?? resolveLayout(existing.kind), desired.layout);
    this.compareLeafable(changes, existing, desired);
    return changes;
  }

  private compareLeafable(
    changes: Record<string, [unknown, unknown]>,
    existing: ExistingSlide,
    desired: SlideAttributes,
  ): void {
    // `kind` is the runtime slide type produced by the Emitter:
    // content | section | image | typst.
    switch (String(existing.kind)) {
      case "content":
        compare(changes, "body", existing.body, desired.body);
        break;
      case "section":
        compare(changes, "body", existing.body, desired.body);
        compare(changes, "theme", existing.theme, desired.theme ?? null);
        break;
      case "image":
        compare(changes, "caption", existing.caption, desired.caption);
        break;
      case "typst":
        compare(changes, "source", existing.source, desired.source);
        break;
    }
  }
}

function compare(changes: Record<string, [unknown, unknown]>, field: string, current: unknown, desired: unknown): void {
  if (String(current ?? "").trim() === String(desired ?? "").trim()) return;
  changes[field] = [current ?? null, desired ?? null];
}

/** Re-export so callers can check a layout name without importing layouts. */
export { layoutKey };